text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,194 @@
+- Start Date: 2018-11-25
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Ember Data Record Identifiers 
+
+## Summary
+
+A follow-up RFC to [RecordData]() that simplifies APIs and ensures capabilities that were intended by the original RFC but not delivered.
+
+Specifically it ensures that `RecordData` can be implemented as a `singleton`, eliminates several redundant APIs that creeped into the original implementation, and simplifies the method signatures of `RecordData` APIs.
+
+The primary shift is that we would now utilize `json-api` resource identifiers and upgrade them within the store to `RecordIdentifier`s to communicate `type` `id` and `lid` information between APIs. We also deprecate use of `clientId` in favor of `lid`.
+
+## Motivation
+
+- nested saves / transactions
+- websockets
+- reload
+- better cache serialization
+- simpler cache and relationship management
+
+## Detailed design
+
+### New Typings
+
+```typescript
+import { Value as JSONValue } from 'json-typescript';
+
+export type Meta = { [k: string]: JSONValue };
+
+export interface ResourceIdentifier {
+  lid?: string;
+  id: string | null;
+  type: string;
+  meta?: Meta;
+}
+
+export interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
+}
+```
+
+### New APIs.
+
+* **recordIdentifierFor**
+
+`import { recordIdentifierFor } from '@ember-data/store';`
+
+```typescript
+/*
+  Returns the RecordIdentifier for the given ResourceIdentifier.
```

Comment:
Do we need a split between a `recordIdentifier` and a `resourceIdentifier` if we are not guaranteeing stability? If you don't have to pass the given `recordIdentifier`, this just seems like wasted roundtrip

Reply:
I think you resolved the difference for yourself below :) one is non-normalized (things we accept from the API), the other is the stricter normalized version (e.g. MUST have `lid` etc.)

Reply:
I would be more comfortable with `ResourceIdentifier` being exported from a JSONAPI-specific package, and not at all part of ember-data core.

Reply:
@dgeb I considered that but no available lib matches what we actually expect.

Specifically:

We allow null id, and lid is not in spec yet. We also will have to handle migrating off of clientId to favor lid.

I can see a future in which this doesn’t need to be provided by ember-data though.

Reply:
@runspired @dgeb can we not make that json-api specific package as part of this change?  presumably with the intent to eventually (or perhaps immediately) move it into the `json-api` org?